keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+experiencing a transitioning state) is high. The effect of a single node
+experiencing a state change increases the queue of updates to be performed for
+all service load balancers, which can lead to:
+
+- rate-limiting of the KCCM by the cloud provider
+
+- or worse: a non-trivial amount of time taken to sync all services in the
+  cluster, leading to a de-synchronized state between the Kubernetes cluster
+  (its state-of-the-world for what concerns services) and the cloud provider.
+
+Besides the problem listed above there is also an important issue arising from
+updating the service spec in certain cases. An example of such is: updating the
+service field `externalTrafficPolicy` from `Cluster` to `Local` or vice versa,
+which will also impact ingress traffic. Specifically, the time taken to update
+the HC and have the service proxy answer accurately on all nodes, will have the
+load balancer drop incoming traffic. This change can take up to minutes to get
+applied completely and thus has important impacts on the service SLAs linked to
+the ingress of the service.
+
+## Motivation
+
+The motivation of this KEP is to design a mechanism that allows for the load
+balancer set to change without re-configuring the load balancer when any
+__existing__ nodes transition in/out of the set. Emphasis is being placed on
+"existing", since nodes which are being fully deleted or newly added to the
+cluster will still require a re-configuration (following what was detailed in
+[summary](#summary)) of the load balancer by the KCCM.
+
+To be able to fully address this problem, we will also need to discuss a
+question which remains un-answered, namely: what should the "inclusion
+predicates" be, for any node to get added to / removed from the set?
+
+Addressing and answering the problem presented above will allow the Kubernetes
+cluster to react much more efficiently to transient errors, for what application
+load balancers are concerned. This will increase application ingress SLAs and
+de-couple it from the "level of entropy" impacting circumstantial components /
+constructs which might be orthogonal to the performance of the application
+itself.  
+
+### Goals
+
+#### Defining "healthy"
+
+- Answer the question: what constitutes a "healthy" node from the point of view
+  of an __application load balancer__? Under which conditions should a node be /
+  not be considered as an acceptable next hop for load balancer traffic? Should
+  a healthy __node__ even be the guiding criteria determining where an
+  __application__ load balancer should send traffic?
+
+- Improve the performance of keeping the cloud provider load balancers in sync
+  with the state of the Kubernetes cluster, by dynamically having the load
+  balancer identify the set of nodes to be used, solely through the utilization
+  of their health check (HC) probes.  
+
+#### Align health checks between services
+
+- Reduce service downtime caused by updating fields on the service object which
```

Comment:
> Maybe worth re-evaluating? 

I think so. Most providers should be able to do L7 load balancing and support paths by now. 

I am not willing to add it to this enhancement proposal (since that is probably a much larger debate which would slow this entire KEP down severely). And we need this KEP as to advance on the decision making for the implementation and finalize what was started in [1.25](https://github.com/kubernetes/kubernetes/pull/109706). 

But I could open a new KEP where we could tackle that problem.